File: 20.valid-parentheses.java
import java.util.Stack;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {

    //字符串替换
    public boolean isValid1(String s) {
        if (s.length() % 2 != 0) return false;
        int length = 0;
        while(length != s.length()) {
            length = s.length();
            s = s.replace("()", "");
            s = s.replace("[]", "");
            s = s.replace("{}", "");

        }
        return length == 0;
    }

    public boolean isValid2(String s) {
        if (s.length() % 2 != 0) return false;
        char[] charArray = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < charArray.length; ++i) {
            if (charArray[i] == '[') {
                stack.push(']');
            } else if (charArray[i] == '(') {
                stack.push(')');
            } else if (charArray[i] == '{') {
                stack.push('}');
            } else if (stack.isEmpty() || stack.pop() != charArray[i]) {
                return false;
            }
        }
        return stack.isEmpty();
    }
}
// @lc code=end

File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#
# https://leetcode.com/problems/valid-parentheses/description/
#
# algorithms
# Easy (38.33%)
# Likes:    4506
# Dislikes: 206
# Total Accepted:    923.3K
# Total Submissions: 2.4M
# Testcase Example:  '"()"'
#
# Given a string containing just the characters '(', ')', '{', '}', '[' and
# ']', determine if the input string is valid.
# 
# An input string is valid if:
# 
# 
# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
# 
# 
# Note that an empty string is also considered valid.
# 
# Example 1:
# 
# 
# Input: "()"
# Output: true
# 
# 
# Example 2:
# 
# 
# Input: "()[]{}"
# Output: true
# 
# 
# Example 3:
# 
# 
# Input: "(]"
# Output: false
# 
# 
# Example 4:
# 
# 
# Input: "([)]"
# Output: false
# 
# 
# Example 5:
# 
# 
# Input: "{[]}"
# Output: true
# 
# 
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        # 加个占位，避免pop时indexError
        dict = {'(':')', '[':']', '{':'}', '$':'$'}
        ans = ['$']
        for char in s:
            if char in dict:
                ans.append(char)
            else:
                if dict[ans.pop()] != char: return False
        return len(ans) == 1

        
# @lc code=end

